### PR TITLE
ENH:  Process XRefStm

### LINKED ISSUE #1273: TypeError: 'NumberObject' object is not subscriptable
See https://github.com/py-pdf/PyPDF2/issues/1269 for further details.

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
Linux-5.4.0-122-generic-x86_64-with-glibc2.29

$ python -c "import PyPDF2;print(PyPDF2.__version__)"
2.10.3
```
## Code + PDF

This is a minimal, complete example that shows the issue:

```python
import PyPDF2
with open("shiv_resume.pdf", "rb") as f:
  pdfreader = PyPDF2.PdfFileReader(f, strict=False)
```

PDF used above: [shiv_resume.pdf](https://github.com/shivpatankar/shivpatankar.github.io/blob/121f0889cfa1eec9096722fb08f30cf4f94358cf/media/resume/shiv_resume.pdf)

## Traceback

This is the complete Traceback I see:

```
Xref table not zero-indexed. ID numbers for objects will be corrected.
Xref table not zero-indexed. ID numbers for objects will be corrected.
Superfluous whitespace found in object header b'17' b'23'

Traceback (most recent call last):
  File "test4.py", line 3, in <module>
    pdfreader = PyPDF2.PdfFileReader(f, strict=True)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 1775, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 275, in __init__
    self.read(stream)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 1279, in read
    self._read_xref_tables_and_trailers(stream, startxref, xref_issue_nr)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 1435, in _read_xref_tables_and_trailers
    xrefstream = self._read_pdf15_xref_stream(stream)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 1515, in _read_pdf15_xref_stream
    assert cast(str, xrefstream["/Type"]) == "/XRef"
TypeError: 'NumberObject' object is not subscriptable
```

### LINKED ISSUE #1279: TypeError: 'NoneType' object is not iterable
See https://github.com/py-pdf/PyPDF2/issues/1269 for further details.

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
Linux-5.4.0-122-generic-x86_64-with-glibc2.29

$ python -c "import PyPDF2;print(PyPDF2.__version__)"
2.10.3
```
## Code + PDF

This is a minimal, complete example that shows the issue:

```python
import PyPDF2
with open("pca_var.pdf", "rb") as f:
  pdfreader = PyPDF2.PdfFileReader(f, strict=False)
  metadata = pdfreader.metadata.copy() if pdfreader.metadata is not None else {}
```

PDF used above: [pca_var.pdf](https://github.com/tnd79bk/EbookMLCB/blob/2441571df00c59c57b431c74ef4000faaa795a52/Chapters/figs/07_DimemsionalityReduction/27_pca/latex/pca_var.pdf)

## Traceback

This is the complete Traceback I see:

```
Object 26 0 not defined.

Traceback (most recent call last):
  File "test.py", line 4, in <module>
    metadata = pdfreader.metadata.copy() if pdfreader.metadata is not None else {}
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 329, in metadata
    retval.update(obj)  # type: ignore
TypeError: 'NoneType' object is not iterable
```

### LINKED ISSUE #1292: ValueError: invalid literal for int() with base 10: b\'0001.24.30\' in _read_standard_xref_table
See https://github.com/py-pdf/PyPDF2/issues/1269 for further details, this reports another issue I've come accross.

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
Linux-5.4.0-122-generic-x86_64-with-glibc2.29

$ python -c "import PyPDF2;print(PyPDF2.__version__)"
2.10.3
```
## Code + PDF

This is a minimal, complete example that shows the issue:

```python
import PyPDF2
with open("BreezeManual.pdf", "rb") as f:
  pdfreader = PyPDF2.PdfFileReader(f, strict=False)
  full_content = " ".join([page.extractText() for page in pdfreader.pages])
```

PDF used above: [BreezeManual.pdf](https://github.com/wise2c-devops/breeze/blob/83d7604462d4ebdee0a9ed5fcc957e6780591993/manual/BreezeManual.pdf)

## Traceback

This is the complete Traceback I see:

``` plain
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 1775, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 275, in __init__
    self.read(stream)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 1279, in read
    self._read_xref_tables_and_trailers(stream, startxref, xref_issue_nr)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 1427, in _read_xref_tables_and_trailers
    startxref = self._read_xref(stream)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 1449, in _read_xref
    self._read_standard_xref_table(stream)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 1391, in _read_standard_xref_table
    offset, generation = int(offset_b), int(generation_b)
ValueError: invalid literal for int() with base 10: b'0001.24.30'
```

**The PDF can be read using a normal PDF viewer.**

### LINKED ISSUE #1294: KeyError: b'1' in read_string_from_stream
See https://github.com/py-pdf/PyPDF2/issues/1269 for further details, this reports another issue I've come accross.

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
Linux-5.4.0-122-generic-x86_64-with-glibc2.29

$ python -c "import PyPDF2;print(PyPDF2.__version__)"
2.10.3
```
## Code + PDF

This is a minimal, complete example that shows the issue:

```python
import PyPDF2
with open("AN EXACT ANALYTICAL SOLUTION OF KEPLER'S EQUATION.pdf", "rb") as f:
  pdfreader = PyPDF2.PdfFileReader(f, strict=False)
  full_content = " ".join([page.extractText() for page in pdfreader.pages])
```

PDF used above: [AN EXACT ANALYTICAL SOLUTION OF KEPLER'S EQUATION.pdf](https://github.com/timedegree/playground_files/blob/1db71cc545a8f487f0257145438cc016523b46c4/%E8%AE%BA%E6%96%87/AN%20EXACT%20ANALYTICAL%20SOLUTION%20OF%20KEPLER%27S%20EQUATION.pdf)

Another example triggering the exact same issue: [How to time-stamp a digital document-88624977e595442fde2b0bd4fe3a0fae.pdf](https://github.com/behfarkhosravi/behfarkhosravi.github.io/blob/391b879169d21716636158857751379af093eb7c/assets/files/How%20to%20time-stamp%20a%20digital%20document-88624977e595442fde2b0bd4fe3a0fae.pdf)

## Traceback

This is the complete Traceback I see:

``` plain
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/generic/_utils.py", line 80, in read_string_from_stream
    tok = escape_dict[tok]
KeyError: b'1'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_page.py", line 1538, in extractText
    return self.extract_text()
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_page.py", line 1510, in extract_text
    return self._extract_text(
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_page.py", line 1157, in _extract_text
    content = ContentStream(content, pdf, "bytes")
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/generic/_data_structures.py", line 689, in __init__
    self.__parse_content_stream(stream_bytes)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/generic/_data_structures.py", line 719, in __parse_content_stream
    operands.append(read_object(stream, None, self.forced_encoding))
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/generic/_data_structures.py", line 830, in read_object
    return read_string_from_stream(stream, forced_encoding)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/generic/_utils.py", line 95, in read_string_from_stream
    tok = b_(chr(int(tok, base=8)))
ValueError: invalid literal for int() with base 8: b'169'
```

**The PDF can be read using a normal PDF viewer.**

### LINKED ISSUE #1295: AttributeError: 'NoneType' object has no attribute 'get_object'
See https://github.com/py-pdf/PyPDF2/issues/1269 for further details, this reports another issue I've come accross.

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
Linux-5.4.0-122-generic-x86_64-with-glibc2.29

$ python -c "import PyPDF2;print(PyPDF2.__version__)"
2.10.3
```
## Code + PDF

This is a minimal, complete example that shows the issue:

```python
import PyPDF2
with open("Segmentation & Activation Lab.pdf", "rb") as f:
  pdfreader = PyPDF2.PdfFileReader(f, strict=False)
  full_content = " ".join([page.extractText() for page in pdfreader.pages])
```

PDF used above: [Segmentation & Activation Lab.pdf](https://github.com/eknee/dep/blob/32bb9045f2c23c5db05ab9ac3e76d4a8e97655da/Lab%20Guides/Day%204%20-%20Activating/Segmentation%20%26%20Activation%20Lab.pdf)

## Traceback

This is the complete Traceback I see:

``` plain
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_page.py", line 1538, in extractText
    return self.extract_text()
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_page.py", line 1510, in extract_text
    return self._extract_text(
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_page.py", line 1146, in _extract_text
    cmaps[f] = build_char_map(f, space_width, obj)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_cmap.py", line 21, in build_char_map
    encoding, space_code = parse_encoding(ft, space_code)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_cmap.py", line 124, in parse_encoding
    enc: Union(str, DictionaryObject) = ft["/Encoding"].get_object()  # type: ignore
AttributeError: 'NoneType' object has no attribute 'get_object'

```

**The PDF can be read using a normal PDF viewer** and the PDF even comes from Adobe.

Another example:
- [Day 4 - Segmentation & Activation.pdf](https://github.com/eknee/dep/blob/74b8cd412d97533c6c607d285fa0cf5e6322f088/Lab%20Guides/Day%204%20-%20Activating/Day%204%20-%20Segmentation%20%26%20Activation.pdf)

### PR DESCRIPTION
**Pull request by pubpub-zz**
Fixes #1273
Fixes #1279 
Fixes #1292
Fixes #1294 
Fixes #1295

ROB: Cope with xref starting on \r\n
ROB: Escaped octal code followed by decimal int
ROB: Cope with some corrupted entries in xref table
ROB: Extend xref autorepair cases

### CODE REVIEW

**Comment by pubpub-zz** [comment]
@MartinThoma,
Ready for review

**Comment by pubpub-zz** [comment]
stdby

File: tests/test_xmp.py [hunk 1]
```diff
@@ -172,10 +172,10 @@ def test_dc_subject():
 def test_issue585():
     url = "https://github.com/mstamy2/PyPDF2/files/5536984/test.pdf"
     name = "mstamy2-5536984.pdf"
-    reader = PdfReader(BytesIO(get_pdf_from_url(url, name=name)))
     with pytest.raises(PdfReadError) as exc:
+        reader = PdfReader(BytesIO(get_pdf_from_url(url, name=name)))
         reader.xmp_metadata
-    assert exc.value.args[0].startswith("XML in XmpInformation was invalid")
+    assert exc.value.args[0].startswith("Stream length not defined")
```
Comment (MartinThoma) [inline tests/test_xmp.py:178]: Why did this change? I guess the `reader.xmp_metadata` isn't even touched, is it?

Comment (MartinThoma) [reply tests/test_xmp.py:178]: Before this PR, one could at least get the number of pages:

```
assert len(reader.pages) == 5
```

I guess with this PR it no longer works?

Comment (pubpub-zz) [reply tests/test_xmp.py:178]: I had to modify the test result. I did not analyze further

Comment (pubpub-zz) [reply tests/test_xmp.py:178]: > Before this PR, one could at least get the number of pages:
> 
> ```
> assert len(reader.pages) == 5
> ```
> 
> I guess with this PR it no longer works?

under analysis

Comment (pubpub-zz) [reply tests/test_xmp.py]: The PDF was corrupted : the XRef package had a /Length key corrupted. I've changed the code to discard the loading of the XRef object to allow the main program to recover to a maximum  information : you can now get the metadata 😊
the access to number of pages is (still?) possible

**Comment by pubpub-zz** [comment]
I had to merge iss_1292 to have a global PR.

this PR is now complete

File: tests/test_xmp.py [hunk 2]
```diff
@@ -172,8 +172,8 @@ def test_dc_subject():
 def test_issue585():
     url = "https://github.com/mstamy2/PyPDF2/files/5536984/test.pdf"
     name = "mstamy2-5536984.pdf"
-    reader = PdfReader(BytesIO(get_pdf_from_url(url, name=name)))
     with pytest.raises(PdfReadError) as exc:
+        reader = PdfReader(BytesIO(get_pdf_from_url(url, name=name)))
```
Comment (MartinThoma) [inline tests/test_xmp.py:176]:
```suggestion
    reader = PdfReader(BytesIO(get_pdf_from_url(url, name=name)))
    with pytest.raises(PdfReadError) as exc:
```

Comment (MartinThoma) [reply tests/test_xmp.py:176]: @pubpub-zz What was the reason to move this? I've just tested it locally; it seems not to be necessary to change the test

File: tests/test_reader.py [hunk 3]
```diff
@@ -1113,13 +1115,24 @@ def test_corrupted_xref_table():
...
     url = "https://github.com/py-pdf/PyPDF2/files/9444748/BreezeManual.failed.pdf"
     name = "BreezeMan2.pdf"
     reader = PdfReader(BytesIO(get_pdf_from_url(url, name=name)))
-    try:
-        reader.pages[0].extract_text()
-    except Exception:
-        pass  # Exception normal
-    else:
-        raise Exception("page 0 should not be corrupted")
+
```
Comment (MartinThoma) [inline tests/test_reader.py:1122]:
```suggestion
    reader.pages[0].extract_text()
```

File: tests/test_reader.py [hunk 4]
```diff
@@ -1113,13 +1115,24 @@ def test_corrupted_xref_table():
     name = "BreezeMan1.pdf"
     reader = PdfReader(BytesIO(get_pdf_from_url(url, name=name)))
     reader.pages[0].extract_text()
-
+    # slightly different
```
Comment (MartinThoma) [inline tests/test_reader.py:1118]:
```suggestion

```

File: tests/test_xmp.py (hunk 2 above)
Comment (pubpub-zz) [reply tests/test_xmp.py:176]: I prefer to leave appart and clear any exception of PdfReader init and just focus on the exception raised by the extraction of metadata

File: tests/test_reader.py [hunk 5]
```diff
@@ -1113,13 +1115,23 @@ def test_corrupted_xref_table():
...
     name = "BreezeMan2.pdf"
     reader = PdfReader(BytesIO(get_pdf_from_url(url, name=name)))
-    try:
-        reader.pages[0].extract_text()
-    except Exception:
-        pass  # Exception normal
-    else:
-        raise Exception("page 0 should not be corrupted")
+    reader.pages[0].extract_text()
+
```
Comment (MartinThoma) [inline tests/test_reader.py:1122]:
```suggestion


```

**Comment by pubpub-zz** [comment]
5 sec before me 😝

**Comment by MartinThoma** [comment]
I'll look into applying black automatically in the CI as an extra commit today :smile: 

Also, I want to make flake8 run in parallel to the tests and mypy after pytest so that I can still see issues there in a failed run.

File: tests/test_xmp.py (hunk 2 above)
Comment (MartinThoma) [reply tests/test_xmp.py:176]: Let me check again; I thought there is no exception during init. If there is, then the second line is not necessary and should be removed (the `reader.xmp_metadata`)

**Comment by pubpub-zz** [comment]
I don't think it worth it.
the line missing came from the code review.
One thing I've noticed is that 3.10 check is performed twice. Do you know why ?(for energy saving)

File: tests/test_xmp.py (hunk 2 above)
Comment (MartinThoma) [reply tests/test_xmp.py:176]: I've just ran it again. The exception happens in the `reader.xmp_metadata` line and thus only that line should be guarded by the `with pytest.raises ...` block. This way it is clear that it's that line which causes the exception.

**Comment by MartinThoma** [comment]
> One thing I've noticed is that 3.10 check is performed twice.

It's a different test scenario. `pycryptodome` is removed in that test run.

**Pull request merged**